Sources_v3/StreamChatUI/ChatChannel/ChatVC.swift

```diff
@@ -71,15 +71,26 @@ open class ChatVC<ExtraData: ExtraDataTypes>: ViewController,
             let duration = notification.userInfo?[UIResponder.keyboardAnimationDurationUserInfoKey] as? Double,
             let curve = notification.userInfo![UIResponder.keyboardAnimationCurveUserInfoKey] as? UInt
         else { return }
-        
-        messageComposerBottomConstraint?.constant = -frame.height
+
+        let localFrame = view.convert(frame, from: nil)
+        messageComposerBottomConstraint?.constant = -(view.bounds.height - localFrame.minY)
+
+        let collectionDelta = view.bounds.height - view.safeAreaInsets.bottom - localFrame.minY
+        let needUpdateContentOffset = !messageList.collectionView.isDecelerating && !messageList.collectionView.isDragging
+        let newContentOffset = CGPoint(
+            x: 0,
+            y: messageList.collectionView.contentOffset.y + collectionDelta
+        )
```

Comment:
1. We updating message composer constraint, for it to be on top of keyboard
```
messageComposerBottomConstraint?.constant = -(view.bounds.height - localFrame.minY)
```
2. We calculate how much we need to update collection view content offset, so bottom message still visible, when keyboard appears
```
let collectionDelta = view.bounds.height - view.safeAreaInsets.bottom - localFrame.minY
let newContentOffset = CGPoint(
    x: 0,
    y: messageList.collectionView.contentOffset.y + collectionDelta
)
```
3. But we want to touch content offset if collection view in idle state. Otherwise scroll will be canceled.
```
let needUpdateContentOffset = !messageList.collectionView.isDecelerating && !messageList.collectionView.isDragging
```

Reply:
Thanks. I must say I wouldn't mind having these comments directly in the code.